lib/tbot/ca_rotation.go

```diff
@@ -0,0 +1,217 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tbot
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+	libUtils "github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// See https://github.com/gravitational/teleport/blob/1aa38f4bc56997ba13b26a1ef1b4da7a3a078930/lib/auth/rotate.go#L135
+// for server side details of how a CA rotation takes place.
+//
+// We can leverage the existing renewal system to fetch new certificates and
+// CAs.
+//
+// We need to force a renewal for the following transitions:
+// - Init -> Update Clients: So we can receive a set of certificates issued by
+//   the new CA, and trust both the old and new CA.
+// - Update Clients, Update Servers -> Rollback: So we can receive a set of
+//   certificates issued by the old CA, and stop trusting the new CA.
+// - Update Servers -> Standby: So we can stop trusting the old CA.
+
+// debouncer accepts a duration, and a function. When `attempt` is called on
+// debouncer, it waits the duration, ignoring further attempts during this
+// period, before calling the provided function.
+//
+// This allows us to handle multiple events arriving within a short period, and
+// attempts to reduce the risk of the server going away during a renewal by
+// deferring the renewal until after the server-side elements of the rotation
+// have occurred.
+type debouncer struct {
+	mu    sync.Mutex
+	timer *time.Timer
+
+	// debouncePeriod is the amount of time that debouncer should wait from an
+	// initial trigger before triggering `f`, and in that time ignore further
+	// attempts.
+	debouncePeriod time.Duration
+
+	// f is the function that should be called by the debouncer.
+	f func()
+}
+
+func (rd *debouncer) attempt() {
+	rd.mu.Lock()
+	defer rd.mu.Unlock()
+
+	if rd.timer != nil {
+		return
+	}
+
+	rd.timer = time.AfterFunc(rd.debouncePeriod, func() {
+		rd.mu.Lock()
+		defer rd.mu.Unlock()
+		rd.timer = nil
+
+		rd.f()
+	})
+}
+
+const caWatchRetryLimit = 15
+
+// caRotationLoop continually triggers `watchCARotations` until the context is
+// cancelled. This allows the watcher to be re-established if an error occurs.
+//
+// caRotationLoop also manages debouncing the renewals across multiple watch
+// attempts.
+func (b *Bot) caRotationLoop(ctx context.Context) error {
+	rd := debouncer{
+		f: func() {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			select {
+			case b.reloadChan <- struct{}{}:
+				b.log.Infof("Renewal triggered due to CA rotation.")
+			default:
+				b.log.Debugf("Renewal already queued, ignoring reload request.")
+			}
+		},
+		debouncePeriod: time.Second * 10,
+	}
+	jitter := libUtils.NewJitter()
+
+	for ctx.Err() == nil {
+		for attempt := 1; attempt <= caWatchRetryLimit; attempt++ {
```

Comment:
Hmm. I may just not fully understand the flow here, but I wonder if this is really something that should have a retry limit. Would it be terribly inappropriate to keep trying forever (with sensible backoffs)?

As I understand the outer loop we'll only currently refresh the attempt counter when the watcher finishes on its own. If a bot is running for a long time (I'd argue weeks/months of uptime is reasonable) it seems like we might eventually stop handling CA rotations due to accumulated errors even if we were happily listening for events 99% of the time.

Alternatively, maybe we could reset the event counter on some condition, or if we do want it to have a limit, it might be good to pretty loudly log that we'll no longer respond to CA events for the remainder of the session.

Reply:
From what I can recall, my original intention with this behaviour was to crash out after enough failed attempts, and defer to the superviser (systemd etc). However, looking at the code, that's not what I've implemented here at all. I think with the current implementation, it will go through a series of back offs, before resetting the back off and trying again with the backoff back to a minimal amount, which doesn't seem right. I'll rectify this.